### PR TITLE
Add an option to disable net472 target for local development

### PR DESCRIPTION
**Pull request by safesparrow**
Some contributors disable the net472 target to improve IDE performance when working on the FCS codebase.
As far as I know there is currently no easy way to enable that without modifying the project files, which are in source control.
@auduchinok Do you have any information on the potential speedup and issues for developing?

I propose the following change, which makes it easier to remove the `net472` target for local development without making any changes that would need to be reverted before checking in:
* Make `TargetFrameworks` in the `FSharp.Compiler.Service.sln` conditional
* `net472` is not added if the `_DisableNetFrameworkTarget` MSBuild property is set to `true`
* Automatically import `/Directory.Build.props.user` if it exists

## How to use the flag
The flag is disabled by default, the default behaviour is unchanged
When the user wants to disable `net472` target, they can do one of the following:
1. Create `/Directory.Build.props.user` file with the following content:
```
<Project>

  <PropertyGroup>
    <_DisableNetFrameworkTarget>true</_DisableNetFrameworkTarget>
  </PropertyGroup>

</Project>
```
2. Modify IDE settings and set that property in the IDE.

Option 1. affects both the IDE and any commandline builds.
Option 2. affects IDE.

## How to set the property in Rider
Rider has an option to set custom MSBuild properties for the solution.
To do that use the below page:
image

## Scope
Currently the conditional `TargetFrameworks` is only applied in projects within the `FSharp.Compiler.Service.sln`.
If we are going to allow this setting, I think it's safer to expand it to all multi-targetting projects.

## Potential issues
When using this optimisation, I can think of these potential issues:
1. `dotnet build`/`Build.cmd` failures due to missing `net472` target - I'm not sure how the build scripts work in this regard, and how often users want to check that everything works on both targets.
2. Unintended use of some APIs that are not supported in `net472` that would otherwise be detected earlier when multitargetting - is that possible?

### CODE REVIEW

**Comment by safesparrow** [comment]
@auduchinok as I understand you do this for local development (disable one of the target frameworks) - apologies if I'm mistaken.

If you do, would you be able to give some insight to the following:
- What exactly do you change?
- What do you think the impact is?
- Any issues you had with it.
- Is there a nicer way to do this?

**Comment by auduchinok** [comment]
> @auduchinok as I understand you do this for local development (disable one of the target frameworks) - apologies if I'm mistaken.

Yes, it makes R# not analyze the second target framework, so features that need the type check results don't have to wait for it. It makes editing considerably faster in Rider, and I can't name any downsides.

**Comment by auduchinok** [comment]
Maybe we could have a property like `IsServiceSolution` (checking the solution name with `'$(SolutionName)' == 'FSharp.Compiler.Service'`) in the Directory.Build.props? We could use it directly in the project files or define `DisableNetFrameworkTarget` property using it.

**Comment by vzarytovskii** [comment]
I think, having a user property file is a good thing, we probably shouldn't do it implicitly (based on the solution name) though.

**Comment by auduchinok** [comment]
@vzarytovskii I think it would be good to make the FCS solution as lightweight to work with as possible (that's what the reason to add it in the first place). If there's no need to have the second framework in it, I'd suggest doing it implicitly, so more people would have it this way automatically, especially those who don't know it would make a difference.

**Comment by vzarytovskii** [comment]
> @vzarytovskii I think it would be good to make the FCS solution as lightweight to work with as possible (that's what the reason to add it in the first place). If there's no need to have the second framework in it, I'd suggest doing it implicitly, so more people would have it this way automatically, especially those who don't know it would make a difference.

I'm a bit torn. On one hand, I don't like implicit surprises in msbuild :). But on the other hand, CI builds a full solution, and if anything is wrong, it will catch it.

**Comment by safesparrow** [comment]
My thoughts/questions:
- currently the PR only applies conditional TargetFramework to projects in the FCS solution; shall we expand it to all projects?
- I don't think applying it only to the `FSharp.Compiler.Service.sln` is optimal - I think the decision to optimise or not should be based on "Do I need to test netframework build compatibility before raising a PR?" not "Am I modifying something outside of the FCS solution?" and thus I think is not limited to that solution.

**Comment by vzarytovskii** [comment]
> My thoughts/questions:
> - currently the PR only applies conditional TargetFramework to projects in the FCS solution; shall we expand it to all projects?

I would say, we should, for now, have it on FCS solution and see if it will be needed on others.

> - I don't think applying it only to the `FSharp.Compiler.Service.sln` is optimal - I think the decision to optimise or not should be based on "Do I need to test netframework build compatibility before raising a PR?" not "Am I modifying something outside of the FCS solution?" and thus I think is not limited to that solution. Also commandline build doesn't necessarily specify the solution (in which case it's empty - but maybe that's fine)

By command line build - do you mean "dotnet build" or our buildscripts? In case of former, you'll need to provide the solution. The former is implicit - it builds VS solution by default on Windows,  FSHarp.sln on Linux and macOS or on windows if "noVisualStudio" is specified.

> - There might be other MSBuild tweaks that we could consider with the same goal in mind - they could live in the same `Directory.Build.props.user`

Yep, that's the reason I like this more than implicitly setting the flag.


Let's start with what you have in PR, and see if it works and needs to cover other solutions/projects (likely only FSharp.sln, since VS one has to build projects for net472).

**Comment by safesparrow** [comment]
>By command line build - do you mean "dotnet build" or our buildscripts? In case of former, you'll need to provide the solution. The former is implicit - it builds VS solution by default on Windows, FSHarp.sln on Linux and macOS or on windows if "noVisualStudio" is specified.

`dotnet build` accepts project files too (and can infer a project file), that's when there would be no solution name as I understand (because there is no way to know which solution it should pick).

**Comment by vzarytovskii** [comment]
> >By command line build - do you mean "dotnet build" or our buildscripts? In case of former, you'll need to provide the solution. The former is implicit - it builds VS solution by default on Windows, FSHarp.sln on Linux and macOS or on windows if "noVisualStudio" is specified.
> 
> `dotnet build` accepts project files too (and can infer a project file), that's when there would be no solution name as I understand (because there is no way to know which solution it should pick).

Yeah, it will ask the user to specify solution/project file if there are several choices.

**Comment by safesparrow** [comment]
I wasnt clear. My worry was that if we use solution name in the condition, it might surprise users that `dotnet build FSharp.Compiler.Service.fsproj` doesn't trigger the optimisation.

But this doesn't apply to the user props file approach, so let's ignore that for now.

@vzarytovskii Will all the solutions, including the VS one, build fine with this enabled? I don't have VS installed so I use the `-noVisualStudio` flag and haven't tested this.

@auduchinok I agree there is a benefit to having some optimisations on by default. I would say adding option is a nice first step and we could either make it a default or add some scripts that bootstrap this file/set those props as environment variables so it's easily discoverable (in docs) and easy to use.

I'm off until next week so won't be able to make changes - happy for this to be edited and merged in the form people are happy with.

**Comment by vzarytovskii** [comment]
> I wasnt clear. My worry was that if we use solution name in the condition, it might surprise users that `dotnet build FSharp.Compiler.Service.fsproj` doesn't trigger the optimisation.
> 

Oh, got it. Agree.


> @vzarytovskii Will all the solutions, including the VS one, build fine with this enabled? I don't have VS installed so I use the `-noVisualStudio` flag and haven't tested this.
> 

I am not sure about this. I think VS solution may fail to build, since its "default" TFM is net472. I will try to checkout your branch and test it later this week.

**Comment by auduchinok** [comment]
> @auduchinok I agree there is a benefit to having some optimisations on by default. I would say adding option is a nice first step and we could either make it a default or add some scripts that bootstrap this file/set those props as environment variables so it's easily discoverable (in docs) and easy to use.

I tend to think that unless it's the default, the most people won't notice it at all, since not many people read the docs or are willing to experiment with options like this.

**Comment by vzarytovskii** [comment]
> As for the Disable Net framework option. I would probably reject that, because nice as it is ... and it is super nice ... we can waste days trying to figure out why stuff doesn't work on the CI whilst it works great on our machines. So a self inflicted environment difference is something we try to avoid.

I would say - it's nice to have it as an option, via the flag, but not enabled by default.

**Comment by auduchinok** [comment]
> I would say - it's nice to have it as an option, via the flag, but not enabled by default.

What's the point of targeting .NET Framework in FCS solution, unlike, say, VisualFSharp? Are we aware of anyone actually using this output when working with that solution?

**Comment by safesparrow** [comment]
@KevinRansom 
>normally I would approve it

>As for the Disable Net framework option. I would probably reject that

Not sure then what would be normally approved - the above is the main part of the PR 🙂

My thoughts on the idea of local changes to the build process are:
- many people, including myself, use the officially supported `-noVisualStudio` option, which skips a big chunk of the build and that's an accepted risk
- many people, including myself, unload projects in the solution, because it's just a performance-reliability trade-off many are willing to make
- I would argue that this sort of local change is rather easy to spot, especially for people who make the change to user settings in the first place
- if really needed, we could add an option to the build that ignores any local settings eg. `-ignoreLocalSettings`.
- as @auduchinok points out, we could choose the "don't build net472 in FCS solution only" therefore making it consistent between local and CI environments
- some people, like myself and @auduchinok , already apply the same optimisation, but do it in a more time-consuming way
- in general I feel like dev experience in the codebase could be prioritised more. I understand the "local vs CI inconsistency", but I think we should try and workaround any potential issues

**Comment by KevinRansom** [comment]
I meant the props thing, i mistyped, sorry.

**Comment by safesparrow** [comment]
@KevinRansom If the reason for not skipping the net472 target is to not diverge local and CI environments then the user props file is by definition undesired as its purpose is to customise local environment without affecting the CI.

For me both are desired as they present a reasonable trade-off between performance and reliability.

**Comment by psfinaki** [comment]
In one project I worked, we used a different approach:
1) Developers could locally create a (pre-gitignored) file called ".targetframeworks" and write there the desired list of frameworks to compile for (e.g. "netcore3.1;net6.0")
2) Then in the repo-root Common.props, we had this code:
```xml
  <!-- target framework override -->
  <PropertyGroup Condition="'$(TargetFramework)' == '' and Exists('$(MSBuildThisFileDirectory)\..\.targetframeworks')">
    <TargetFrameworks>$([System.IO.File]::ReadAllText('$(MSBuildThisFileDirectory)\..\.targetframeworks').Split(';'))</TargetFrameworks>
  </PropertyGroup>
```
That was basically it and it usually worked.

But the approach here is also viable, just a bit more verbose IMO.

**Comment by vzarytovskii** [comment]
> In one project where I worked, we used a different approach:
> 
> 1. Developers could locally create a (pre-gitignored) file called ".targetframeworks" and write there the desired list of frameworks to compile for (e.g. "netcore3.1;net6.0")
> 2. Then in the repo-root Common.props, we had this code:
> 
> ```
>   
>   <PropertyGroup Condition="'$(TargetFramework)' == '' and Exists('$(MSBuildThisFileDirectory)\..\.targetframeworks')">
>     <TargetFrameworks>$([System.IO.File]::ReadAllText('$(MSBuildThisFileDirectory)\..\.targetframeworks').Split(';'))</TargetFrameworks>
>   </PropertyGroup>
> ```
> 
> That was basically it and it usually worked.
> 
> But the approach here is also viable, just a bit more verbose IMO.

That won't, likely, work here, since many projects have different sets of frameworks.

**Comment by vzarytovskii** [comment]
> @KevinRansom Kevin Ransom FTE If the reason for not skipping the net472 target is to not diverge local and CI environments then the user props file is by definition undesired as its sole purpose is to customise local environment without affecting the CI.
> 
> For me both are desired as they present a reasonable trade-off between performance and reliability.

There are few concerns about the approach:
1. We don't want it to be **implicit** (e.g. based on the solution name), since it is more _magic_ in our build process, which is already confusing.
2. Using the user file is fine with me (I know Kevin is not very fond of that, but can live with that, from what I understood :)), however, there are two consequences:
	1. If user does `git clean -xdf`, file will be gone (I personally do it at least few times a day). 
	2. Right now, in this PR, if you create the file, and run `Build.cmd` with no `-noVisualStudio` flag, it will fail to build. But this can be fixed.

**TL;DR:** I am personally fine with merging it and see what would UX be with this change, but we potentially will need to make some adjustments to it in the future.

**Comment by safesparrow** [comment]
1. We could have a committed and pre-ignored props file that sets it to false and the user can change the value to true. Ie. `git update-index --skip-worktree Directory.Build.props.user`. This would make it much easier to use the file (change one string rather than create the file from scratch). I'm not sure what `git clean ...` would do with such a file - would it revert any local changes?

2. I would be interested to know why there is a need to run `git clean ...` so often. Is the build lacking a robust `clean` target?

**Comment by safesparrow** [comment]
>Wouldn't be then too easy for user to commit it by accident, changing default value? Or will it be commited with default values, and then ignored (even from git commit -a)?

Hmm, I wrongly assumed the git index command applies to the remote repository - instead it just affects the local index, so it doesn't make much sense.

**Comment by NinoFloris** [comment]
> Or will it be committed with default values, and then ignored (even from git commit -a)?

This is how it works!

**Comment by vzarytovskii** [comment]
> > Or will it be committed with default values, and then ignored (even from git commit -a)?
> 
> This is how it works!

Hm, if we can have this file with default setting, and then ignore it, preventing anyone to accidentally commit it, I guess it will work.

**Comment by safesparrow** [comment]
> > Or will it be committed with default values, and then ignored (even from git commit -a)?
> 
> This is how it works!

It is, but it has to be done by the developer on the local repository - it cannot be automatically applied to all future clones.

**Comment by safesparrow** [comment]
Some options:
1. Forget about the user props file
2. Keep the current version which has to be recreated after `git clean`
3. Suggest people create `../Directory.Build.props` in the parent of the repo.
4. Keep the current version but edit the docs and replace `git clone ...` with `git clone ... -e *.user` - this avoids having to recreate the file.
5. Create `Directory.Build.props.user.baseline` in the repo which people copy to the ignored `Directory.Build.props.user` after a checkout and modify

**Comment by vzarytovskii** [comment]
> 2. Keep the current version which has to be recreated after `git clean`
I think we should go with this one.

> 6. Suggest people use IDE-specific MSBuild properties as described in the PR (doesn't apply to the build script, not sure if VS has this option)

Additionally, can this be part of the Rider's dotsettings for FCS solution @auduchinok?

**Comment by auduchinok** [comment]
> Additionally, can this be part of the Rider's dotsettings for FCS solution @auduchinok?

Interesting idea! Yes, we can add the following to the [settings file](https://github.com/dotnet/fsharp/blob/main/FSharp.Compiler.Service.sln.DotSettings):
```
<s:String x:Key="/Default/Environment/Hierarchy/Build/SolBuilderDuo/CustomGlobalProperties/=TestProperty/@EntryIndexedValue">TheValue</s:String>
```

**Comment by vzarytovskii** [comment]
> EDIT: Although this is then the implicit behaviour we didn't want.
 
I was not very comfortable with the idea of turning it implicitly globally for everything msbuild-related (i.e. in the properties). I think it's okay to have it for FCS in Rider.

**Comment by safesparrow** [comment]
@KevinRansom Now that https://github.com/dotnet/fsharp/issues/13318 is closed, is everything ready to disable the `net472` target? Is there an existing issue for that?

**Comment by KevinRansom** [comment]
Thanks for doing this and prodding us to actually do the work to make CompilerService netstandard work correctly on both desktop and coreclr.

This was achieved by these PR's here:

- https://github.com/dotnet/fsharp/pull/13532
- https://github.com/dotnet/fsharp/pull/13630
- https://github.com/dotnet/fsharp/pull/13603
- https://github.com/dotnet/fsharp/pull/13589
- https://github.com/dotnet/fsharp/pull/13579
- https://github.com/dotnet/fsharp/pull/13578
- https://github.com/dotnet/fsharp/pull/13580
